File: Day - 1/code/03_Implemented_tree.java
import java.util.ArrayList;
import java.util.List;

class TreeNode {
    String data;
    List<TreeNode> children;

    TreeNode(String data) {
        this.data = data;
        this.children = new ArrayList<>();
    }

    void addChild(TreeNode child) {
        children.add(child);
    }
}

class Tree {
    TreeNode root;

    Tree(TreeNode root) {
        this.root = root;
    }

    void printTree(TreeNode node, int level) {
        if (node == null) return;
        System.out.println(" ".repeat(level * 2) + node.data);
        for (TreeNode child : node.children) {
            printTree(child, level + 1);
        }
    }

    void printTree() {
        printTree(root, 0);
    }

    TreeNode find(String data, TreeNode node) {
        if (node == null) return null;
        if (node.data.equals(data)) return node;
        for (TreeNode child : node.children) {
            TreeNode result = find(data, child);
            if (result != null) return result;
        }
        return null;
    }

    TreeNode find(String data) {
        return find(data, root);
    }

    int getHeight(TreeNode node) {
        if (node.children.isEmpty()) return 1;
        int maxHeight = 0;
        for (TreeNode child : node.children) {
            maxHeight = Math.max(maxHeight, getHeight(child));
        }
        return 1 + maxHeight;
    }

    int getHeight() {
        return getHeight(root);
    }

    int countNodes(TreeNode node) {
        int count = 1;
        for (TreeNode child : node.children) {
            count += countNodes(child);
        }
        return count;
    }

    int countNodes() {
        return countNodes(root);
    }

    void printList(TreeNode node) {
        System.out.print(node.data + " ");
        for (TreeNode child : node.children) {
            printList(child);
        }
    }

    void printList() {
        printList(root);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode("Electronics");

        TreeNode laptop = new TreeNode("Laptop");
        laptop.addChild(new TreeNode("Mac"));
        laptop.addChild(new TreeNode("Dell"));

        TreeNode phone = new TreeNode("Phone");
        phone.addChild(new TreeNode("iPhone"));
        phone.addChild(new TreeNode("Samsung"));

        TreeNode tv = new TreeNode("TV");
        tv.addChild(new TreeNode("Sony"));
        tv.addChild(new TreeNode("LG"));

        root.addChild(laptop);
        root.addChild(phone);
        root.addChild(tv);

        Tree tree = new Tree(root);

        System.out.println("Tree structure:");
        tree.printTree();

        System.out.println("\nSearch for 'iPhone':");
        TreeNode found = tree.find("iPhone");
        System.out.println("Found: " + (found != null ? found.data : "Not found"));

        System.out.println("\nTree height: " + tree.getHeight());
        System.out.println("Total nodes: " + tree.countNodes());

        System.out.print("Tree values in sequence: ");
        tree.printList();
    }
}

File: Day - 1/code/03_Implemented_tree.py
# TreeNode class
class TreeNode:
    def __init__(self, data):
        self.data = data
        self.children = []

    def add_child(self, child_node):
        self.children.append(child_node)


# Tree class
class Tree:
    def __init__(self, root):
        self.root = root

    def print_tree(self, node=None, level=0):
        if node is None:
            node = self.root
        print(" " * (level * 2) + str(node.data))
        for child in node.children:
            self.print_tree(child, level + 1)

    def find(self, data, node=None):
        if node is None:
            node = self.root
        if node.data == data:
            return node
        for child in node.children:
            result = self.find(data, child)
            if result:
                return result
        return None

    def get_height(self, node=None):
        if node is None:
            node = self.root
        if not node.children:
            return 1
        return 1 + max(self.get_height(child) for child in node.children)

    def count_nodes(self, node=None):
        if node is None:
            node = self.root
        count = 1
        for child in node.children:
            count += self.count_nodes(child)
        return count

    def print_list(self, node=None):
        if node is None:
            node = self.root
        print(node.data, end=" ")
        for child in node.children:
            self.print_list(child)



if __name__ == "__main__":
    # Create tree
    root = TreeNode("Electronics")

    laptop = TreeNode("Laptop")
    laptop.add_child(TreeNode("Mac"))
    laptop.add_child(TreeNode("Dell"))

    phone = TreeNode("Phone")
    phone.add_child(TreeNode("iPhone"))
    phone.add_child(TreeNode("Samsung"))

    tv = TreeNode("TV")
    tv.add_child(TreeNode("Sony"))
    tv.add_child(TreeNode("LG"))

    root.add_child(laptop)
    root.add_child(phone)
    root.add_child(tv)

    tree = Tree(root)

    # Function Tests
    print("Tree structure:")
    tree.print_tree()
    
    print("\n\nSearch for 'iPhone':")
    found = tree.find("iPhone")
    print("Found:", found.data if found else "Not found")

    print("\nTree height:", tree.get_height())
    print("Total nodes:", tree.count_nodes())

    print("Tree values in sequence:")
    tree.print_list()
